src/bson.c

```diff
@@ -1201,6 +1289,195 @@ static bool php_phongo_bson_state_parse_type(zval* options, const char* name, ph
+}
+
+static php_phongo_field_path_map_element* field_path_map_element_alloc(void)
+{
+	php_phongo_field_path_map_element* tmp = ecalloc(1, sizeof(php_phongo_field_path_map_element));
+
+	tmp->entry = php_phongo_field_path_alloc(true);
+
+	return tmp;
+}
+
+bool php_phongo_bson_state_add_field_path(php_phongo_bson_typemap* map, char* field_path_original, php_phongo_bson_typemap_types type, zend_class_entry* ce TSRMLS_DC)
+{
+	char*                              saveptr;
+	char*                              field_path_str = estrdup(field_path_original);
+	const char*                        element;
+	php_phongo_field_path_map_element* field_path_map_element;
+
+	field_path_map_element = field_path_map_element_alloc();
+
+	/* Loop over all the segments. A segment is delimited by a "." */
+	element = php_strtok_r(field_path_str, ".", &saveptr);
+	while (element) {
+		php_phongo_field_path_push(field_path_map_element->entry, element, PHONGO_FIELD_PATH_ITEM_NONE);
+		element = php_strtok_r(NULL, ".", &saveptr);
+	}
+
+	efree(field_path_str);
+
+	field_path_map_element_set_info(field_path_map_element, type, ce TSRMLS_CC);
+	map_add_field_path_element(map, field_path_map_element);
+
+	return true;
+}
+
+static void field_path_map_element_dtor(php_phongo_field_path_map_element* element)
+{
+	php_phongo_field_path_free(element->entry);
+	efree(element);
+}
+
+void php_phongo_bson_typemap_dtor(php_phongo_bson_typemap* map)
+{
+	size_t i;
+
+	for (i = 0; i < map->field_paths.size; i++) {
+		field_path_map_element_dtor(map->field_paths.map[i]);
+	}
+	efree(map->field_paths.map);
+
+	map->field_paths.map = NULL;
+}
+
+/* Loops over each element in the fieldPaths array (if exists, and is an
+ * array), and then checks whether each element is a valid type mapping */
+bool php_phongo_bson_state_parse_fieldpaths(zval* typemap, php_phongo_bson_typemap* map TSRMLS_DC) /* {{{ */
+{
+	zval*      fieldpaths = NULL;
+	HashTable* ht_data;
+
+	if (!php_array_existsc(typemap, "fieldPaths")) {
+		return true;
+	}
+
+	fieldpaths = php_array_fetchc_array(typemap, "fieldPaths");
+
+	if (!fieldpaths) {
+		phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "The 'fieldPaths' element is not an array");
+		return false;
+	}
+
+	ht_data = HASH_OF(fieldpaths);
+
+#if PHP_VERSION_ID >= 70000
+	{
+		zend_string* string_key = NULL;
+		zend_ulong   num_key    = 0;
+		zval*        property;
+
+		ZEND_HASH_FOREACH_KEY_VAL(ht_data, num_key, string_key, property)
+		{
+			zend_class_entry*             map_ce = NULL;
+			php_phongo_bson_typemap_types map_type;
+
+			if (!string_key) {
+				phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "The 'fieldPaths' element is not an associative array");
+				return false;
+			}
+
+			if (!php_phongo_bson_state_parse_type(fieldpaths, ZSTR_VAL(string_key), &map_type, &map_ce TSRMLS_CC)) {
+				return false;
+			}
+
+			if (!php_phongo_bson_state_add_field_path(map, ZSTR_VAL(string_key), map_type, map_ce TSRMLS_CC)) {
```

Comment:
`php_phongo_bson_state_add_field_path()` always returns true. In the event a failure case is introduced, I suppose we'll want to ensure it throws an exception. Correct?

Is that why you're preemptively passing in `TSRMLS_CC` here (consistency with `php_phongo_bson_state_parse_type()`)?

Reply:
I'd wish to say that I added ``TSRMLS_CC`` because of that, but that's not true :-) But yes, perhaps we should leave it as long as the strtok loop can create failure notifications with the three cases that you suggested.

Reply:
Aye, let's revisit this if you decide to add exceptions for tokenization errors.

Reply:
Which I've now just done :-).